### PR TITLE
fix(versioning): refuse a version restore whose child history was pruned (sc-120012)

### PR DESCRIPTION
**Pull request by mikebridge**
### SUMMARY

Restoring a `SqlaTable` version reconstructs its columns/metrics from the child shadow rows valid at the target transaction. Version-history retention can have legitimately pruned exactly those rows while the parent's row survives — and SQLAlchemy-Continuum's reverter then finds no child version at the target and **deletes the live child outright**: a durable write of an incomplete column/metric set, silently reported as success (sc-120012, surfaced by the independent review of #44012).

**Restore now fails closed.** Before any write, the engine verifies the target's child history and refuses with a user-facing 422 (`PrunedChildHistoryError`: "…can no longer be fully restored: … pruned by version-history retention. The entity was left unchanged.") when a needed row was pruned.

Detection is validity-**interval** analysis over the surviving rows — the pruner also deletes the `version_transaction` rows and cascades the change records, so the chain is the only evidence:

- a surviving row's `[transaction_id, end_transaction_id)` covering the target proves the state either way (non-DELETE: present, restored; DELETE: provably absent — a closed end is Continuum's evidence of a re-insert);
- with no covering interval, a last same-parent row at/before the target that is DELETE is accepted as absence;
- if no row is at/before the target, an earliest surviving INSERT is accepted as born-after evidence; erased earlier history means it need not be the original birth;
- anything else is a gap that may have covered the target: refuse.

The verification read is taken `FOR UPDATE` in the transaction that owns the restore — every row the verdict depends on, covering DELETE witnesses included (the reverter selects the latest surviving row at/before the target *then* excludes DELETEs, so a pruned absence witness would resurrect an older incarnation) — closing the check-to-write window under READ COMMITTED: the pruner's SERIALIZABLE pass blocks on the locks until the restore commits. On SQLite, where `FOR UPDATE` is a no-op and pysqlite's legacy mode holds no transaction for SELECTs, the verifier takes a `BEGIN IMMEDIATE` write reservation up front (issued through the SQLAlchemy connection so contention translates and maps to 422).

Also per the ticket: `get_version`'s parent + child reads now run in ONE snapshot (`REPEATABLE READ` on MySQL/Postgres; an explicit BEGIN on SQLite via the documented pysqlite recipe) so a concurrent prune cannot age out child rows between the two reads, and the deferred notes in `get_version` / `restore.py` carry the sc-120012 disposition.

**Known limitation** (ratified on sc-120012): the guard refuses every *detectable* pruning of needed child history and protects all restores targeting versions within the retention window. Three indistinguishable missing-history shapes can fail open for targets older than the retention window. Preserving their dependency closure is the separate **post-GA** sc-120945 work, not implemented by this PR:

1. a column/metric deleted **after** the target whose entire shadow chain — including its closing DELETE row — has aged out leaves no surviving evidence anywhere (shadow, transaction, and change-record rows are pruned together; change records are per-tx diffs, never a full set);
2. a same-parent re-birth whose window `[X, e)` was retention-pruned via `e` while `X` survived is **observationally identical** to the benign, common case of a purged foreign incarnation of a recycled id, and is accepted as absence. Two invariants ground that acceptance: retention cannot erase the closer of a *surviving* closed DELETE (the pruner deletes rows whose close-transaction is pruned, so the DELETE row would have gone with it), and purge never touches the live parent's own rows — so refusing here would instead permanently brick restores of any live dataset whose recycled child ids ever passed through a later-purged dataset.

3. erased birth and covering rows can leave only a later re-insertion INSERT, indistinguishable from a child genuinely born after the target.

The guard's accepted algebra and residuals are characterized in unit tests; the entirely erased chain also has a DB-level forced-erasure test. These do not prove a retention policy: actual-prune preservation requirements belong to sc-120945. Every refusal WARNING-logs the child's surviving `(tx, end, op)` chain so an unprovable state is diagnosable from logs alone.

Review: three adversarial cross-model rounds (committer + sqlalchemy lenses) folded before opening — including an interval-semantics rewrite of the verifier after a confirmed counterexample, the DELETE-witness lock scope, and the SQLite reservation path.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

API behavior. Before: `POST /api/v1/dataset/<uuid>/versions/<version>/restore` against a version whose child history was pruned → 200 OK, dataset silently loses columns. After: 422 with the message above; the dataset is untouched.

### TESTING INSTRUCTIONS

- `pytest tests/unit_tests/versioning/test_restore.py` — 14-case interval-algebra parametrization including exact create-transaction equality (covered / DELETE-covered / born-after pass; every detectable hole refuses, including the review round's confirmed counterexample), engine guard branches, endpoint 422 mapping pin.
- `pytest tests/integration_tests/versioning/restore_fails_closed_tests.py` — DB-level: command-level refusal with a COMMITTED prune and no test-owned rollback (the command's transactional cleanup owns the unchanged state); refusal-precedes-any-write ordering pin (write phase never entered, no pending ORM entity changes before rollback; the deliberate Core shadow DELETE is not represented by session.deleted); the guard-removed CONTROL proving the incomplete write reappears; intact-history restore unchanged; FOR UPDATE lock test (PG/MySQL: concurrent DELETE hits the dialect's lock-wait error, succeeds after release); SQLite write-reservation pins (real transaction held; concurrent writer blocked; contention surfaces as the command failure type); the documented-limitation pin and a post-foreign-closer restore after that foreign incarnation is erased.

Remediation verification on `7797be4e014d21e82cfc724fd430ab31882ac5e1`: 180 versioning unit tests passed; isolated SQLite integration 9 passed / 1 PostgreSQL-MySQL locking test skipped; applicable full-branch pre-commit passed. Mutating the inclusive comparison to strict `<` made the new boundary case fail, then restoring it returned the suite to green. Independent different-model source review found no publication blocker (three documentation nits remain); PostgreSQL/MySQL locking was not rerun locally. Current-head CI is tracked separately.

### ADDITIONAL INFORMATION
- [x] Has associated issue: sc-120012
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01JRLEJS4mUqKBoPjSjviKUW

### CODE REVIEW

**Review by aminghadersohi — COMMENTED** [review]
The refusal predicate reads sound in both directions; two notes — a boundary case the suite leaves unpinned, and an undercount in the ratified fail-open list.

File: tests/unit_tests/versioning/test_restore.py [hunk 1]
```diff
@@ -131,3 +131,205 @@ def test_single_flush_scope_skips_flush_on_exception() -> None:
...
+
+    assert _child_state_provable_at(rows, target_tx) is expected, case
+
+
+@pytest.mark.parametrize(
+    ("rows", "target_tx", "expected", "case"),
+    [
+        # A surviving non-DELETE row covers the target: complete.
+        ([_Row(5, None, _INSERT)], 10, True, "live row covers"),
+        ([_Row(5, 20, _UPDATE)], 10, True, "closed row covers"),
```
Comment (aminghadersohi) [inline tests/unit_tests/versioning/test_restore.py:235]: No case pins `transaction_id == target_tx`. Changing the covering test's `<=` to `<` passes all 24 unit and 9 integration tests, yet flips a column edited in the very transaction being restored to from restorable to refused — the guard's most common false-positive direction.

```suggestion
        # A surviving non-DELETE row covers the target: complete.
        ([_Row(5, None, _INSERT)], 10, True, "live row covers"),
        ([_Row(5, 20, _UPDATE)], 10, True, "closed row covers"),
        # Boundary: a row created AT the target tx covers it (half-open
        # [tx, end)) — the child changed in the transaction restored to.
        ([_Row(10, None, _UPDATE)], 10, True, "row created at target covers"),
```

File: superset/versioning/restore.py [hunk 2]
```diff
@@ -71,6 +72,194 @@
...
+    and the earliest is the child's birth INSERT (born after). See
+    :func:`_child_state_provable_at` for the interval semantics.
+
+    Anything else means a pruned row MAY have covered ``target_tx`` —
+    fail closed (sc-120012).
+
+    Known limitation (ratified, sc-120012): the fail-closed guard refuses
+    every DETECTABLE pruning of needed child history, including a pruned
+    closed row whose successor survives, and protects all restores
+    targeting versions within the retention window. One residual fails
```
Comment (aminghadersohi) [inline superset/versioning/restore.py]: "One residual" undercounts — born-after fails open too. A same-parent re-birth INSERT can survive while the birth and covering rows prune via their close-tx, leaving a row set identical to the legitimate born-after case, so no read-side check separates them. No fence: the wording is yours.

File: tests/unit_tests/versioning/test_restore.py (hunk 1 above)
Comment (mikebridge) [reply tests/unit_tests/versioning/test_restore.py:235]: Good catch—addressed in 7797be4e014d21e82cfc724fd430ab31882ac5e1 with the exact UPDATE row created at the target transaction. The real negative control (`<=` changed to `<`) failed that case; restoring the inclusive predicate passed all 180 versioning unit tests. Isolated SQLite integration: 9 passed, 1 server-dialect locking test skipped. Current-head CI is still running.

File: superset/versioning/restore.py (hunk 2 above)
Comment (mikebridge) [reply superset/versioning/restore.py]: Good catch—7797be4e014d21e82cfc724fd430ab31882ac5e1 documents all three indistinguishable shapes, including a surviving re-insertion INSERT after erased birth/covering history, and adds a characterization pin. It also strengthens foreign-ID coverage to a target after the closer, with the foreign incarnation removed. The ratified guard is unchanged; dependency-closure retention stays separate and post-GA in sc-120945. The PR summary has been aligned with that limitation.

**Comment by mikebridge** [comment]
Follow-up to the Bito notes in https://github.com/apache/superset/pull/44251#issuecomment-5668401610 and https://github.com/apache/superset/pull/44251#issuecomment-5671326859: 7797be4e014d21e82cfc724fd430ab31882ac5e1 reuses the canonical INSERT/DELETE constants, documents the test helpers and hoists their import, corrects the refusal assertion's scope to ORM changes (not the deliberate Core DELETE), and removes the obsolete foreign-witness log comment. The born-after indistinguishability is documented, not claimed solved: retention preservation remains post-GA sc-120945. Verification: 180 versioning unit tests passed; SQLite integration 9 passed / 1 server-dialect test skipped; full branch changed-file pre-commit passed. Independent source review found no publication blocker, with three documentation nits retained. New-head CI is pending, so this is not an all-green or all-feedback-resolved claim.

CI update: the new-head integration and Docker-init jobs fail before tests with Alembic 'Multiple head revisions'. I verified the PR merge snapshot and master 7b6dd7597c4af49f7f0571b7b33dd247f382218b have the same two heads, 0884f655c0a6 and e2f3a1b9c640, and no migration-file difference. The branch itself has one head. The existing upstream repair is #44364; this PR does not add a duplicate merge migration. CI is blocked, not green.

Repair follow-up (2026-09-17 UTC): #44364 is merged at 4c77b0b924a9bd770110915eacd68f5619e8241d. Published content-identical CI refresh 1a67a8cd1a5986a0db05929a00d5630a2de054db by fast-forward (tree unchanged). The new PR merge snapshot 52b777a69fe6821f420b0a8758cbace67ceec723 includes repaired master and has one Alembic head, 60f94cd6cd11. New integration run https://github.com/apache/superset/actions/runs/35166031034 and unit run https://github.com/apache/superset/actions/runs/35166031028 target this new event, not a rerun of the old broken merge SHA. Fresh local versioning tests: 180 passed; isolated SQLite integration: 9 passed / 1 server-dialect skip; all applicable changed-file pre-commit hooks passed. Remote CI is still running, not claimed green. No guard/retention policy change or duplicate migration; sc-120945 stays post-GA.

File: superset/versioning/restore.py [hunk 3]
```diff
@@ -71,6 +73,205 @@
...
+                shadow.c.end_transaction_id,
+                shadow.c.operation_type,
+            )
+            .where(shadow.c.table_id == entity.id)
+            .order_by(shadow.c.id, shadow.c.transaction_id)
+            .with_for_update()
+        ).all()
+        by_child: dict[int, list[Any]] = {}
+        for row in rows:
+            by_child.setdefault(row.id, []).append(row)
```
[the remark this answers is not in the document]
Comment (mikebridge) [reply superset/versioning/restore.py:193]: Thanks—this is the explicitly ratified fully-erased-history residual, not a case this read-side guard claims to recover. At 1a67a8cd1a5986a0db05929a00d5630a2de054db, restore.py documents all three indistinguishable shapes and test_documented_limitation_fully_pruned_deleted_child_fails_open pins this one. Preserving the missing evidence is the separate post-GA sc-120945 dependency-closure retention work; this PR keeps the accepted guard unchanged.

File: superset/versioning/restore.py [hunk 4]
```diff
@@ -71,6 +73,205 @@
...
+    for row in rows:
+        if row.transaction_id <= target_tx and (
+            row.end_transaction_id is None or row.end_transaction_id > target_tx
+        ):
+            return True
+    at_or_before: list[Any] = [row for row in rows if row.transaction_id <= target_tx]
+    if not at_or_before:
+        return rows[0].operation_type == OPERATION_INSERT
+    last: Any = max(at_or_before, key=lambda row: row.transaction_id)
+    return last.operation_type == OPERATION_DELETE
```
[the remark this answers is not in the document]
Comment (mikebridge) [reply superset/versioning/restore.py:273]: Thanks—the pruned same-parent re-birth window is also an explicitly ratified residual, documented in _child_state_provable_at and pinned by the existing characterization tests. Requiring the DELETE interval to cover the target would reverse the accepted foreign-ID/purge behavior without recovering erased evidence. The retention-side preservation work remains separate and post-GA in sc-120945; this PR does not claim to close that limitation.

**Review by aminghadersohi — COMMENTED** [review]
Both round-1 findings are closed by measurement: `<=`→`<` on the covering predicate now fails the new `row created at target covers` case, and the docstring enumerates all three indistinguishable shapes. One order-dependency note inline.

File: superset/versioning/restore.py [hunk 5]
```diff
@@ -71,6 +73,205 @@
...
+    branch cannot distinguish it from the child's original birth.
+    """
+    for row in rows:
+        if row.transaction_id <= target_tx and (
+            row.end_transaction_id is None or row.end_transaction_id > target_tx
+        ):
+            return True
+    at_or_before: list[Any] = [row for row in rows if row.transaction_id <= target_tx]
+    if not at_or_before:
+        return rows[0].operation_type == OPERATION_INSERT
```
Comment (aminghadersohi) [inline superset/versioning/restore.py]: `rows[0]` silently depends on the `order_by` at :188; the sibling branch below defends itself with `max(...)`. Drop `transaction_id` from that `order_by` and all 180 versioning tests still pass, yet `[INSERT@20→25, DELETE@25]` flips born-after from provable to refused.

```suggestion
        earliest: Any = min(rows, key=lambda row: row.transaction_id)
        return earliest.operation_type == OPERATION_INSERT
```

Comment (mikebridge) [reply superset/versioning/restore.py]: Good catch — addressed in b3f7072f606493cb107f556436aec633cf9c90da using the explicit earliest-transaction selection. The production ORDER BY stays intact. Two unsorted-input cases have independent literal expectations: earliest INSERT proves born-after, while a later INSERT must not conceal an earliest UPDATE. Both failed against rows[0] (one false refusal, one false acceptance); a real revert control reproduced both failures, and the restored final file passes 28 tests. No test/helper sorts the inputs. Full changed-file pre-commit, including MyPy/pylint, and final-source independent review passed; new-head CI remains pending. I did not rerun the full 180-test ORDER BY-removal experiment.

**Review by aminghadersohi — COMMENTED** [review]
`rows[0]` is closed: reverting it fails both new cases, and the ORDER-BY-removal run your reply notes as skipped now passes 182/182 — the proof is order-independent, so that mutant is equivalent rather than uncovered. Nothing open from us.

**Review by aminghadersohi — APPROVED** [review]
